zwrap/sample.go

```diff
@@ -98,70 +103,62 @@ func (s *sampler) With(fields ...zap.Field) zap.Logger {
 }
 
 func (s *sampler) Check(lvl zap.Level, msg string) *zap.CheckedMessage {
-	if !s.check(lvl, msg) {
-		return nil
+	cm := s.Logger.Check(lvl, msg)
+	switch lvl {
+	case zap.PanicLevel, zap.FatalLevel:
+		return cm
 	}
-	return zap.NewCheckedMessage(s.Logger, lvl, msg)
+	if s.sampled(msg) {
+		return cm
+	}
+	return nil
 }
 
 func (s *sampler) Log(lvl zap.Level, msg string, fields ...zap.Field) {
-	if s.check(lvl, msg) {
+	if s.Logger.Check(lvl, msg) != nil && s.sampled(msg) {
```

Comment:
Can we special-case panic/fatal logging here as well? Instead of making the sampling behavior dependent on which logger methods are called (`sampler.Panic(msg)` is different from `sampler.Log(zap.PanicLevel, msg`), we should not sample panic or fatal level logs anywhere.

Also, let's use the `OK` method instead of a nil check (same throughout the leveled logging methods).